### PR TITLE
 How about hOCR? #75

### PR DESCRIPTION
**Pull request by ws233**
I've started with this.
1. Changed the proposed implementation a bit to make it safer.
2. Added one unit test. Perhaps more tests should be added. For example, for different images, if they exist. Also a test to call the function before recognize.
3. Also I've not checked setInputImage yet.

### CODE REVIEW

**Comment by ws233** [comment]
1. That's just a warning.
2. I couldnot find that way so far.

**Comment by kevincon** [comment]
This Stack Overflow suggests that if you have that warning, you won't be able to get an app approved by Apple in the App Store: http://stackoverflow.com/questions/25909870/xcode-6-and-embedded-frameworks-only-supported-in-ios8/25910262#25910262 Do you agree? If so, we should absolutely fix it.

A fix I can confirm that works is to require the minimum deployment target to be 8.0. I know you and @BamX said we should support iOS 7.0, but obviously that won't be an option if apps won't be approved while we have this warning. :(

**Comment by BamX** [comment]
The only solution I see is to release 4.0.0 without new framework version for now and than release another version with that framework and iOS 8 target. That allow iOS 7.1+ devices use all new functionality we implemented since 3.1.1. What are you think about it?

**Comment by ws233** [comment]
There are more answers in that topic, which says thay can upload an app to the appstore without any errors or warnings. And those answers are not so old. So perhaps, Apple changed something in their approving to allow this. Also there is a link where Apple says that a definetely bug on their side:
https://devforums.apple.com/message/999579#999579

**Comment by ws233** [comment]
Does anybody have some developer profile to check if the app could be uploaded to app store and pass automatic validation?

File: TestsProject/TestsProjectTests/hOCR.m [hunk 1]
```diff
@@ -0,0 +1,61 @@
+//
+//  hOCR.m
```
Comment (BamX) [inline TestsProject/TestsProjectTests/hOCR.m]: Why are you moved this tests related to RecognitionTests to another file?

Comment (ws233) [inline TestsProject/TestsProjectTests/hOCR.m]: I thought there would be more tests.
Should I move them back?

Comment (BamX) [inline TestsProject/TestsProjectTests/hOCR.m]: It closely related to recognition(it is recognition). You could create ```context``` for hOCR. Don't think that there many tests are needed. Any way it is better them all to lay in one place for now.

Comment (ws233) [inline TestsProject/TestsProjectTests/hOCR.m]: ok.

**Comment by ws233** [comment]
We even do not need to upload the app to the app store to verify it. There is a verify button on the xCode archives page. Unfortunately, I do not have a valid production certificate now, so this validation doesn't work on my computer.
Can anyone check this?

**Comment by kevincon** [comment]
So I created a dummy app in iTunes Connect and was able to run validation on Template Framework Project. Here are the errors:

screen shot 2015-01-06 at 8 48 09 pm

Notice the second error in particular. It confirms that the minimum deployment target for the Tesseract OCR framework must be >= 8.0.

So I propose we make the minimum deployment for the framework to be 8.0. When I make this change, here's the result of validation:

screen shot 2015-01-06 at 9 02 58 pm

And that's just an unrelated issue with the template framework project missing a necessary icon.

So do you guys agree we must make the minimum deployment target for the Tesseract OCR framework to be 8.0? That seems to be the required solution to pass app validation.

**Comment by ws233** [comment]
In such case all the Apps including our framework should also have minimum deployment target iOS7.

**Comment by ws233** [comment]
By the way, what's about your xCode version?

**Comment by ws233** [comment]
I think I've found a decision. If a dynamic framework is not allowed, why do we use it? :)
There is a static framework also available in xCode6.
It operates exactly the same as our manually compiled framework.
I'm trying this, but already have faced an issue.
There is a linker error about double 'main' functions. One is from our test project. The second one is from libtesseract_all.
Do you know why there is a main function in libtesseract_all and why there was no such linker error in AggregateTarget? I watched its build script, but couldnot find anything, which can help me to understand this.

**Comment by ws233** [comment]
Ok. I'll answer my own question. That's due to -all_load flag I've added to the TestProject. That flag forces linker to load all available symbols from the library. And our libtesseract_all.a contains tesseract viewer, which is an app actually with the main function. More info about the viewer is here https://code.google.com/p/tesseract-ocr/wiki/ViewerDebugging. 

Is it really necessary to have this app inside our framework? It seems that Java VM is necessary to run this app. So could it be run under iOS? Especially if it's a part of another app and doesn't have it's own main function.

Unfortunately I'm not familiar with bash scripts, so I cannot forbid libtesseract_viewer.a to link with the other .a files into libtesseract_all.a. Can some one take a look on this and remove it from the building and linking stages of our build_dependencies.sh script?

Regarding static framework. I've successed with it. There is a push request.
Since it's not a dynamic library, there shouldn't be any warning or errors mentioned above.

@kevincon, can you check that this PR can be validated by the Apple in iTunes Connect?

**Comment by kevincon** [comment]
Nice job! The pull request is successfully validated by Apple in iTunes Connect, and I can still install the Cocoa Pod locally with a test project.

I found another warning, though. When you build the TestsProject or try testing the TestsProject, I get this warning:

```
ld: warning: Auto-Linking supplied '/Users/kcon/Library/Developer/Xcode/DerivedData/Tesseract-OCR-iOS-ahivhaxwpuazxlcuoglxnsisyeam/Build/Products/Debug-iphonesimulator/TesseractOCR.framework/TesseractOCR', framework linker option at /Users/kcon/Library/Developer/Xcode/DerivedData/Tesseract-OCR-iOS-ahivhaxwpuazxlcuoglxnsisyeam/Build/Products/Debug-iphonesimulator/TesseractOCR.framework/TesseractOCR is not a dylib
```

Any ideas on how to remove it? I know it's just a warning, but we really shouldn't have any warnings at all in this library.

By the way, the Xcode version I am using is Version 6.1.1 (6A2008a).

**Comment by ws233** [comment]
Nice to hear. :)
I didn't get this warning or just missed it, so let me investigate it a bit.

**Comment by ws233** [comment]
Ah! I found what's wrong.
Forgot to remove TesseractOCR from Dynamic linking section of Test Project. Did this for example project, but forfot for the test one.
Pushed.

**Comment by ws233** [comment]
Should I create a new task to remove libtesseract_viewer.a from the libtesseract_all.a?
Actually, it may cause troubles for some users, who adds -load_all flag to their projects due to some reasons.

**Comment by kevincon** [comment]
Can you provide me with some instructions to reproduce the linker error you're receiving? Where is the -load_all flag?

If I can reproduce the issue, then I can work on modifying the build_dependencies.sh script to leave that lib out of libtesseract_all.a.

**Comment by ws233** [comment]
Here it is.
Open the Build settings iether of the TestsProject or TemplateFramework one. Find Other Linker Flags parameter in the Linker section. Add the flag there to any of the configurations or even both.

**Comment by kevincon** [comment]
I see, so why do we need the all_load flag? The template framework project seems to run okay without it. When you say all_load loads all symbols, do you mean debug symbols that could be used to identify internal parts of upstream Tesseract that might cause errors?

Also, was eef0760 supposed to fix the `ld: warning: Auto-Linking supplied '/Users/kcon/Library/Developer/Xcode/DerivedData/Tesseract-OCR-iOS-ahivhaxwpuazxlcuoglxnsisyeam/Build/Products/Debug-iphonesimulator/TesseractOCR.framework/TesseractOCR', framework linker option at /Users/kcon/Library/Developer/Xcode/DerivedData/Tesseract-OCR-iOS-ahivhaxwpuazxlcuoglxnsisyeam/Build/Products/Debug-iphonesimulator/TesseractOCR.framework/TesseractOCR is not a dylib` warning? I still get it unfortunately when I build the TestsProject target. :(

**Comment by ws233** [comment]
Please, make a clean first. and then rebuild.

**Comment by ws233** [comment]
I think you still have a dynamic lib binary somewhere in the build folder. So after cleaning and rebuilding that warning should disappear. At least it desappearedd on my computer.
And the same code doesnot produce the warning in example project.
So everything should be ok.

**Comment by kevincon** [comment]
I did clean for both the framework target and the TestsProject target and I still get the same warning.

Here's a screenshot that shows I'm on the correct commit and the resulting warning:

screen shot 2015-01-08 at 12 32 35 am

**Comment by ws233** [comment]
I'm really sorry for this. I also did a clean build and there is a warning. So perhaps I forgot something else.

**Comment by kevincon** [comment]
No problem, let me know if you discover what's causing it. I'm going to start looking at removing the viewer from the tesseract lib.

**Comment by ws233** [comment]
Try this commit pls. That helped me https://github.com/bpoplauschi/CrashlyticsFramework/issues/18.
It seems that the framework should also be added to the TestsProjectTests target. Before it was never there.

**Comment by kevincon** [comment]
Nice job! 825c05d fixed the warning for me. :)

I also sent you a pull request (ws233/Tesseract-OCR-iOS#3) that removes the viewer from the upstream tesseract lib and adds `-all_load` to the template framework project and the tests project, and both work successfully now. :)

**Comment by ws233** [comment]
You did ask about the use cases where this flag is necessary.
I could find only few so far.
1. Imagine the user calls some of the methods from the library dynamically in runtime. That means no actual calls [instance method], but  instead calls like [instance performSelector:selectroFromString(@"method")];
In such case there is no static linking with method, therefore it's not in the final executable. So to force linker to link with all functions and classes from the framework -all_load flag shoud be added.
2. Another case is the user tries to use some of the upstream Tesseract functions straightforward without our wrappers. Perhaps, it cannot, because our framework doesnot contain -all_load flag. That means only the functions and classes used in our framework are linked with our framework. So perhaps, we should add this flag and link leptonica and tesseract completely to the framework. What do you think?

**Comment by kevincon** [comment]
I don't think we should add `-add_load` flag to the framework because I like the idea of abstracting away the underlying upstream Tesseract/Leptonica libraries. If accessing the upstream lib functions is the main use case for using `-all_load`, then I don't think it should be added to the template framework project or the tests project either (especially since we don't even use non-API upstream lib functions in those projects).

I think that if the user needs access to more functionality from the upstream libs than we provide through our methods/classes then they should work with us on GitHub issues to request the functionality they need.

Does that sound okay? @BamX, what do you think about this?

**Comment by ws233** [comment]
I'm ok with this. Should I revert the commit?

**Comment by kevincon** [comment]
Rather than revert the commit, can you make a new commit that removes `-all_load` from the framework, the template framework project, and the tests project? The issue is that I had added `-all_load` in my commit that removes the viewer, so it wouldn't be as simple as just reverting the last few commits.

**Comment by kevincon** [comment]
Can you also remove `-all_load` from TestsProject? It's still there.

After that I think we're good to finally merge and close this pull request. :)

**Comment by ws233** [comment]
There is a build error after coverall has been added.
I've searched the internet and found the following:
http://stackoverflow.com/questions/7949781/undefined-symbols-for-architecture-i386-upgrading-project-to-ios-5?answertab=votes#tab-top
the first answer solves the issue, but I would like to understand why it appeared?

@kevincon, do you have any ideas?

**Comment by ws233** [comment]
Actually not. It works on my local computer, but fails on Travis CI.

Perhaps, that's due to load function and changes in applicationWillTerminate, which have been added to the test project, but have not been done to the ExampleProject.
I'll check this.

**Comment by kevincon** [comment]
It looks like your commit that adds `GCC_INSTRUMENT_PROGRAM_FLOW_ARCS = YES;` to Template Framework Project fixes it, but that's not ideal because this means the user of our library will also need to add `GCC_INSTRUMENT_PROGRAM_FLOW_ARCS = YES;` to be able to use TesseractOCR.framework in a debug configuration.

The solution I saw online is to add a new build configuration to TesseractOCR.framework called like "Coverage" that is a duplicate of "Debug", add the `GCC_INSTRUMENT_PROGRAM_FLOW_ARCS = YES;` and also "Generate Test Coverage" flags to the "Coverage" build configuration, and then make sure that the unit tests project uses "Coverage" when building the TesseractOCR.framework.

Do you want to give that a shot?

**Comment by ws233** [comment]
Is it possible not to share this target with the root project?
I mean will the user see this target in their own project if they include it in their workspace?

**Comment by ws233** [comment]
Anyway, let me try this.

**Comment by ws233** [comment]
I think we do not need this. Cause release build for test project is build without any error.
Also there are 2 warning about int, which you've already fixed. Could you merge them back?

**Comment by kevincon** [comment]
We do need it, because users will be trying out our library in Debug configurations as it is the default for build and run in Xcode, so they will get the gconv error unless they somehow know to add the instrumentation flag to their project's build settings, which isn't ideal.

I think I'm getting pretty close to getting it to work, let me keep trying.

**Comment by ws233** [comment]
Yup. I got it already. That's why removed my previous comment :)
Ok. I'm waiting for your changes so.

**Comment by kevincon** [comment]
Okay, check out ws233/Tesseract-OCR-iOS#4 and see what you think. :)

**Comment by kevincon** [comment]
Great, it looks like that worked to fix coveralls support without requiring the user's project to have the instrumentation flag. Are you ready for me to merge this now?

**Comment by ws233** [comment]
yep

**Comment by ws233** [comment]
And then I will upload some unit tests

**Comment by kevincon** [comment]
Merging and closing!

**Pull request merged**